core/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -699,76 +741,135 @@ private void allocateUnassigned() {
+                        updateMinNode = true;
+                    }
+                    if (updateMinNode) {
+                        minNode = node;
+                        minWeight = currentWeight;
+                        decision = currentDecision;
                     }
                 }
             }
             if (decision == null) {
                 // decision was not set and a node was not assigned, so treat it as a NO decision
                 decision = Decision.NO;
             }
-            return Tuple.tuple(decision, minNode);
+            String explanation = null;
+            if (explain) {
+                if (decision.type() == Type.YES) {
+                    assert minNode != null;
+                    explanation = "shard assigned to node [" + minNode.getNodeId() + "]";
+                } else if (decision.type() == Type.THROTTLE) {
+                    assert minNode != null;
+                    explanation = "shard assignment throttled on node [" + minNode.getNodeId() + "]";
+                } else {
+                    explanation = "shard cannot be assigned to any node in the cluster";
+                }
+            }
+            return ShardAllocationDecision.fromDecision(
+                decision,
+                minNode != null ? minNode.getNodeId() : null,
+                explanation,
+                nodeExplanationMap
+            );
+        }
+
+        // can we short circuit execution of the unassigned allocation decision due
+        // all nodes being throttled?
+        private boolean canShortCircuit(Set<ModelNode> throttledNodes) {
+            if (allocation.debugDecision()) {
+                return false;
+            }
+            return throttledNodes.size() >= nodes.size();
+        }
+
+        // can we short circuit execution of the unassigned allocation decision for the
+        // given node?
+        private boolean canShortCircuitShard(ShardRouting shard, ModelNode node, float minWeight, float currentWeight,
+                                             Set<ModelNode> throttledNodes) {
+            if (allocation.debugDecision()) {
+                return false;
+            }
+            return throttledNodes.contains(node) // node is throttled
+                       || node.containsShard(shard) // node already contains a copy of the shard
+                       || currentWeight > minWeight; // moving the shard would not improve the balance
+        }
+
+        // provide an explanation, if in explain mode
+        private String explain(String explanation) {
+            if (allocation.debugDecision()) {
+                return explanation;
+            } else {
+                return null;
+            }
+        }
+
+        // add a node explanation safely (only if the map is not null, meaning we are in explain mode)
+        private static void addNodeExplanation(Map<String, NodeExplanation> map, String nodeId, Decision decision, float weight) {
```

Comment:
this method is used in two places. maybe easier to just do the `if (explain) map.put(nodeId, new NodeExplanation(decision, weight));` there.